contracts/vaults/BalleMaster.sol

```diff
@@ -0,0 +1,349 @@
+// contracts/vaults/BalleMaster.sol
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.3;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "../token/BALLEv2.sol";
+import "../interfaces/IStrategy.sol";
+
+/**
+ * @dev Implementation of the Master of BALLE.
+ * This contract will take care of all rewards calculations and distribution of BALLE tokens in vaults.
+ * It's ownable and the owner is the only who can manage the active vaults and it's parameters for rewards distribution.
+ * The ownership will be transferred to the Governance GNOSIS Safe.
+ */
+contract BalleMaster is Ownable, ReentrancyGuard {
+    using SafeERC20 for IERC20;
+
+    // Info of each user
+    struct UserInfo {
+        uint256 shares; // User shares of the vault.
+        uint256 rewardDebt; // Reward debt. See explanation below.
+        //
+        // We do some fancy math here. Basically, any point in time, the amount of BALLEs
+        // entitled to a user but is pending to be distributed is:
+        //
+        //   amount = user.shares / sharesTotal * lockedTotal
+        //   pending reward = (amount * vault.accBallePerShare) - user.rewardDebt
+        //
+        // Whenever a user deposits or withdraws LP tokens to a vault. Here's what happens:
+        //   1. The vault's `accBallePerShare` (and `lastRewardBlock`) gets updated.
+        //   2. User receives the pending reward sent to his/her address.
```

Comment:
At this point, the user will receive just the reward generated by this user proportionally? Avoiding the trick to get reward withdrawing when there was no liquidity?

Reply:
What do you mean by "when there was no liquidity"??
Each user will get the pending rewards generated from its start to the moment the LP tokens are modified, by a new deposit or a withdrawn, because no two temporal rates could be saved for the user account, only the corresponding rate from las change. Is the same way PancakeSwap or Autofarm works.

Reply:
ok, I was talking about this time when you could retire LP getting rewards and the fees were lower (or harvest), I'm still trying to figure how this can happens and how to avoid it, I remember you suggest avoiding harvest until minimum liquidity was achieved, maybe it's not related to this function?

Reply:
Not related here, this is for BALLE rewards exclusively. What do you refer will be done on strategy and withdrawal fee takes care of that.